File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.py
from collections import deque


def solution(begin, target, words):
    visited = [False] * len(words)
    queue = deque()
    queue.append((begin, 0))

    while queue:
        cur_word, cur_step = queue.popleft()
        if cur_word == target:
            return cur_step

        for i in range(len(words)):
            cnt = 0
            if not visited[i]:
                for j in range(len(words[i])):
                    if words[i][j] != cur_word[j]:
                        cnt += 1
            if cnt == 1:
                visited[i] = True
                queue.append((words[i], cur_step + 1))
    return 0

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    public int solution(String begin, String target, String[] words) {
           boolean[] visited = new boolean[words.length];
        Deque<Current> deque = new ArrayDeque<>();
        deque.add(new Current(begin, 0));

        while (!deque.isEmpty()) {
            Current current = deque.pollFirst();

            if (current.word.equals(target)) {
                return current.step;
            }

            for (int wordIdx = 0; wordIdx < words.length; wordIdx++) {
                if (visited[wordIdx]) {
                    continue;
                }

                int count = 0;
                for (int charIdx = 0; charIdx < words[wordIdx].length(); charIdx++) {
                    if (words[wordIdx].charAt(charIdx) != current.word.charAt(charIdx)) {
                        count++;
                    }
                }

                if (count == 1) {
                    visited[wordIdx] = true;
                    deque.add(new Current(words[wordIdx], current.step + 1));
                }
            }
        }

        return 0;
    }

    private static class Current {
        private final String word;
        private final int step;

        public Current(String word, int step) {
            this.word = word;
            this.step = step;
        }
    }
}